### PR TITLE
M✔ ◾ Fix YouTube upload/metadata honesty (#672, #861, #798, #645)

### LINKED ISSUE #672: 🐛 Bug - Upload shows green tick but no YouTube link when channel not created
[▶️ Watch the video (02:15)](https://www.youtube.com/watch?v=Owf5Ib96z3U)

Hi SSW Yakshaver user

### Describe the Bug
After recording and attempting to upload a video to YouTube, YakShaver shows a green tick and indicates "uploading" but does not provide a link to the uploaded video. The user had connected YouTube, but their Google account did not have a YouTube channel created. YakShaver did not detect the missing channel and the upload failed silently without showing an error.

### To Reproduce
1. Open the YakShaver Desktop app and record a video.
2. Connect a Google/YouTube account that does not yet have a YouTube channel created.
3. Attempt to upload the recorded video to YouTube via the app.
4. Observe that the app shows a green tick or "uploading" success state but no video link appears and no error is shown.

### Expected Behavior
If the connected YouTube account does not have an active channel, the app should detect this and prompt the user to create a channel. If an upload fails, the app should display a clear error message indicating the upload failed and not show a misleading success indication.

### Tasks
- [ ] Investigate
- [ ] Fix

### More Information
User reports they were prompted by YouTube Studio to create a channel during the process. After creating the channel manually, uploads should work, but YakShaver did not check for channel existence during the connect/upload flow. The flow should validate channel existence during YouTube connection and during upload confirmation.

### Environment
 - Device: Desktop
 - Browser: N/A (YakShaver Desktop app)
 - OS: Not specified

### Screenshots

Thanks!

### PR DESCRIPTION
**Pull request by calumjs**
## Summary
A cluster of "the YouTube upload/metadata stage lies about its status" bugs, all in `process-video-handlers` (the `UPLOADING_VIDEO` / `UPDATING_METADATA` stages) and the workflow UI. The unifying theme: **a stage's displayed status now matches what actually happened.**

Closes #672, #861, #798, #645.

## Fixes (one commit each)

**#672 — green upload tick with no link.** `processFileVideo` completed `UPLOADING_VIDEO` unconditionally, even when `uploadVideo` returned `{success:false}` without throwing (e.g. the Google account has **no YouTube channel** → `youtubeSignupRequired`). Now the stage only completes on real success, otherwise it fails with the upload error (workflow continues so the work item is still created). Adds `describeYouTubeUploadError` for actionable no-channel copy + a unit test.

**#861 — metadata fails but UI says "all good".** The Final Result badge was driven solely by the AI's self-reported `Status`, ignoring a failed `UPDATING_METADATA` stage. `FinalResultPanel` now reads the workflow state it already receives and renders a **"Completed with a warning"** banner with the stage error.

**#798 — "Updating Metadata" shown/errors for external links.** Metadata only applies to a video we uploaded and own. For an external YouTube link or a failed/absent upload the stage is now explicitly **skipped** (and `WorkflowStepCard` already hides skipped stages), so it no longer lingers as a stuck/erroring step.

**#645 — metadata step "not ticked".** Both panels render together, so a *successful* metadata stage already ticks green via its `completeStage` broadcast — the "not ticked" report was the **metadata-failure** case, which #861/#798 now render honestly (red + warning, or hidden) instead of an ambiguous stuck step.

## Verification — honest caveats
- `npm run build` (tsc) + UI `tsc --noEmit` + `biome` clean; the #672 detection has a unit test (5 cases, green).
- **Live smoke is limited for this cluster** (unlike #833): #672 needs a Google account with *no* channel, #861/#645 need a real successful YouTube upload that then fails metadata. So these lean on code review + the unit test rather than a clean live repro. Happy to do a YouTube-connected live pass if a suitable account is available.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by calumjs — COMMENTED** [review]
## Review: YouTube metadata honesty (#874)

This PR improves how upload/metadata failures are surfaced, but the central acceptance criteria of #861 ("don't mark the run successful when metadata fails") are not actually met, and the #672 fix is asymmetric (first-run only, not retry). Details below, grouped by severity.

### High

**1. Metadata failure still returns an overall successful workflow result** — `src/backend/ipc/process-video-handlers.ts` (`@@ -646,12 +650,8 @@`, around line 690-702)
On a metadata error the code calls `workflowManager.failStage(UPDATING_METADATA, metadataErrorMsg)`, then falls through to `cleanupTempFiles` and `return { success: true, youtubeResult, mcpResult, workflowId }`. The RetryResult is unconditionally `success: true` despite the failure. This violates #861 AC#3 ("overall process is not marked successful when metadata update fails") and AC#5 ("a bug report is only marked successfully processed when all required steps complete without errors"). Note the persisted status is also unaffected: `useShaveManager.ts` (~line 43-44) only sets `ShaveStatus.Failed` when the LLM `Status === "fail"`, so a metadata failure never flips the persisted/badge status either.
*Remedy:* return a non-success/warning result on metadata failure and drive the persisted ShaveStatus from stage outcomes, not solely the LLM Status field.

**2. #672 "green tick, no link" regression returns on the RETRY path** — `src/backend/services/workflow/workflow-retry-service.ts:145-155` (`retryUploadingVideo`)
The first-run path now routes through `applyUploadStageOutcome(...)`, but the retry path still calls `workflowManager.completeStage(UPLOADING_VIDEO, {...})` unconditionally with no `youtubeResult.success` check. `uploadVideo()` returns `{ success: false }` without throwing for the no-YouTube-channel case, so the `catch` at :164 never fires. Reproduction: Google account with no YouTube channel → first-run upload now correctly fails → the failed card shows a Retry button → user clicks Retry → re-upload returns `{ success: false }` again → `completeStage()` still runs → green upload tick with no link (the exact #672 symptom). The new unit tests cover only the first-run helper, not this branch.
*Remedy:* route the retry upload through the same `applyUploadStageOutcome(youtubeResult, filePath, workflowManager)` helper, and add a retry-path test mirroring the first-run one.

### Medium

**3. Final panel shows a green "Success" badge alongside "Completed with a warning"** — `src/ui/src/components/workflow/FinalResultPanel.tsx` (`@@ -678,6 +683,27 @@` and `@@ -803,6 +829,14 @@`, badge ~line 815/828)
The badge is derived solely from the AI output: `const status = (isJson && parsed?.Status) ...` rendered as `{status && <StatusBadge status={status} />}`. The new stage-failure detection only writes `stageWarning` and feeds the amber banner; it has no coupling to the badge. On the #861 repro (metadata fails while AI emits `Status: "success"`) the panel renders a green "Success" pill AND the warning banner simultaneously — contradicting #861 AC#1/AC#3. No UI test asserts the badge downgrades on a failed stage.
*Remedy:* derive the displayed final status from both the AI status and workflow stage failures (render warning/failed when UPLOADING_VIDEO or UPDATING_METADATA failed).

**4. Stale "Completed with a warning" banner persists after a successful metadata retry** — `src/ui/src/components/workflow/FinalResultPanel.tsx` (`onProgressNeo` block ~line 683-705; `resetForNewRun` ~line 614-625)
`stageWarning` is set in the new block but cleared only in `resetForNewRun`, which fires only on the `isRetryStage` (enter EXECUTING_TASK) or new-recording transitions. A metadata-only retry routes through `retryFromMiddleStage` → `processVideoSource(..., "updating_metadata")`, which skips EXECUTING_TASK, so `isRetryStage` stays false and `resetForNewRun` never runs. Across the retry broadcasts `uploading_video` stays `completed` and `updating_metadata` is `not_started`/`in_progress`/`completed`, so neither the `if (upload failed)` nor `else-if (metadata failed)` branch runs. With no trailing `else`, the stale warning persists even after the metadata retry succeeds, so the panel falsely claims "Completed with a warning" on a now-clean run.
*Remedy:* make the warning derivation total — recompute it from current state on every broadcast and clear it (`else setStageWarning(null)`) when neither stage is failed, or compute it as derived state rather than sticky state.

**5. #861 root cause (description sanitization/validation) is not addressed** — `src/backend/services/video/video-metadata-builder.ts` (no sanitization) + `process-video-handlers.ts:653-696`
#861's root cause is YouTube rejecting the AI-generated description ("invalid video description", typically from disallowed chars/keywords). AC#4 ("validates or sanitizes video descriptions before sending them to the YouTube API") and AC#6 ("the app is updated to comply with the new requirements") are unaddressed. `video-metadata-builder.ts` builds `snippet.description = appendChapters(metadata.description, ...)` straight from the LLM/fallback output untouched, and `updateVideoMetadata` forwards `snippet` to the YouTube API as-is. The metadata update will keep failing on the very inputs that triggered #861 — the user just sees a nicer error. The PR body textually claims "Closes #861" and lists it as fixed, which is misleading (GitHub only auto-closes #672 here, but the body still presents #861 as resolved).
*Remedy:* either sanitize/validate the description before `updateVideoMetadata` (strip/escape rejected chars, enforce length/keyword rules) to satisfy AC#4/AC#6, or explicitly scope this PR to the honesty ACs and de-scope #861 from "Closes" to "partially addresses" so the sanitization work stays tracked.

**Review by calumjs — COMMENTED** [review]
## Review summary

The core fix (no longer showing a green tick at the UPLOADING stage when YouTube has no channel) is solid and well-tested. Three issues remain that affect correctness and spec completeness.

### High

**Description length cap uses characters, but the YouTube API limit is 5000 bytes** — `src/backend/services/video/youtube-metadata-sanitizer.ts`

`truncate(value, max) => value.slice(0, max)` counts UTF-16 code units, and `YOUTUBE_DESCRIPTION_MAX_LENGTH = 5000` is then applied as a character cap. The YouTube Data API documents `snippet.description` as having "a maximum length of **5000 bytes**" (not characters), while `snippet.title` is correctly a 100-**character** cap. (https://developers.google.com/youtube/v3/docs/videos)

For multi-byte UTF-8 content this under-truncates: a description of 3000 `🦬` emoji slices to 5000 UTF-16 code units = 10000 UTF-8 bytes — 2x the API limit — so `videos.update` is still rejected with the exact "invalid video description" error this sanitizer exists to prevent. This is a realistic and anticipated input: the PR's own test deliberately puts `🦬` in a description.

Secondary: slicing on UTF-16 code units can split a surrogate pair at the 5000 boundary, emitting a lone surrogate — another validator-reject risk.

Remedy: truncate the description by UTF-8 byte length (e.g. `TextEncoder` / `Buffer.byteLength`) to `<= 5000` bytes on a code-point boundary; keep title at 100 characters. Optionally use `Array.from()` / `Intl.Segmenter` to avoid splitting surrogate pairs. Add a test with a long emoji/CJK description asserting byte length `<= 5000` — the existing test only asserts `out.description.length === 5000` (UTF-16 length), which passes even when the byte length is double.

### Medium

**#672's connect-time channel validation is not implemented (only upload-time)** — `src/backend/services/auth/youtube-client.ts`

Issue #672's Expected Behavior states: "If the connected YouTube account does not have an active channel, the app should detect this and prompt the user to create a channel," and More Information specifies validation "during YouTube connection **and** during upload confirmation" — two detection points.

This PR only adds the upload-time path: `describeYouTubeUploadError` fires when `uploadVideo` returns `success:false`. The connect flow is unchanged — `authenticate()` still returns `{ success: true, userInfo: userInfo ?? undefined }` unconditionally, and although `getUserInfo()` already calls `channels.list({ mine: true })` and sets `channelName: channel?.snippet?.title || undefined`, nothing acts on its absence. A Google account with no channel is still treated as successfully connected, so the user only discovers the problem after recording and attempting an upload.

Because the PR says "Closes #672" with no deferral note, merging auto-closes the issue with an explicitly-stated AC unmet.

Remedy: after OAuth, treat a missing channel from `channels.list({ mine: true })` as an actionable auth failure/warning that surfaces the "create a channel" prompt at connect time — or explicitly note on the PR/issue that connect-time detection is deferred so #672 isn't auto-closed as fully done.

File: src/backend/services/video/youtube-metadata-sanitizer.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+export function sanitizeYouTubeText(value: string): string {
+  return value
+    .replace(/</g, "‹") // single left-pointing angle quotation mark
+    .replace(/>/g, "›") // single right-pointing angle quotation mark
+    .replace(/\r\n?/g, "\n") // normalise CRLF/CR to LF (keeps a single LF)
+    .replace(CONTROL_CHARS_TO_STRIP, "");
+}
+
+function truncate(value: string, max: number): string {
+  return value.length <= max ? value : value.slice(0, max);
```
Comment (calumjs) [inline src/backend/services/video/youtube-metadata-sanitizer.ts]: `truncate` slices by UTF-16 code units (`value.slice(0, max)`), but the YouTube API caps `snippet.description` at 5000 **bytes**, not characters. A 5000-code-unit emoji/CJK description can be ~2x the byte limit, so `videos.update` still rejects with "invalid video description". Truncate by UTF-8 byte length on a code-point boundary instead (and keep title at 100 chars). See summary review for details.

File: src/backend/services/auth/youtube-client.ts [hunk 2]
```diff
@@ -192,9 +194,11 @@ export class YouTubeClient {
         origin: "upload",
       };
     } catch (error) {
+      // Report for telemetry, but surface user-facing copy — notably the no-channel case (#672).
+      formatAndReportError(error, "youtube_upload");
       return {
         success: false,
-        error: formatAndReportError(error, "youtube_upload"),
+        error: describeYouTubeUploadError(error),
```
Comment (calumjs) [inline src/backend/services/auth/youtube-client.ts:235]: This adds the **upload-time** no-channel handling, but #672 also requires **connection-time** validation ("validate channel existence during YouTube connection AND during upload confirmation"). `authenticate()` still returns `success: true` even when the account has no channel, and `getUserInfo()`'s `channels.list({ mine: true })` result (`channelName` left undefined) is never acted on — so a channel-less account connects "successfully" and the user only discovers the problem here, at upload. Either add a connect-time check that prompts the user to create a channel, or note the deferral so #672 isn't auto-closed as fully done. See summary review.

File: src/backend/services/video/youtube-metadata-sanitizer.ts (hunk 1 above)
Comment (calumjs) [reply src/backend/services/video/youtube-metadata-sanitizer.ts]: Fixed in 9162f9d. The description is now truncated by UTF-8 byte length (5000 bytes) via a new `truncateBytes()` that walks code points (the string iterator yields whole surrogate pairs) and stops before any code point that would exceed the budget — so it never splits a surrogate or emits a lone one. Title stays a 100-character cap via `truncateChars()`. Renamed the constant `YOUTUBE_DESCRIPTION_MAX_LENGTH -> YOUTUBE_DESCRIPTION_MAX_BYTES`. Added tests: a 3000×🦬 description (12000 bytes) now clamps to ≤5000 bytes (1250 emoji), and a surrogate-on-the-boundary case asserts no lone surrogate remains.

File: src/backend/services/auth/youtube-client.ts (hunk 2 above)
Comment (calumjs) [reply src/backend/services/auth/youtube-client.ts:235]: Fixed in 9162f9d. `authenticate()` now validates channel existence at connect time: after OAuth it checks that `channels.list({ mine: true })` returned an item and, if not, clears the half-connected tokens (so the state stays disconnected and the user can cleanly reconnect) and returns `success: false` with actionable copy — "This Google account doesn't have a YouTube channel yet. Create one at youtube.com, then reconnect." That surfaces through the existing auth-error UI path, so a channel-less account no longer connects "successfully". Presence is keyed off the `items` list rather than the title, so a titleless channel isn't misread as missing. Added `youtube-client.test.ts` covering the no-channel (fails + clears tokens), has-channel, and titleless-channel cases. This satisfies the connection-time half of #672 alongside the existing upload-time handling.

**Review by calumjs — COMMENTED** [review]
## Review summary

One confirmed finding. No high-severity blockers; one medium maintainability/DRY issue worth addressing before merge.

### Medium

**Duplicated "required post-creation stage failed" rule (DRY / AGENTS.md Rule 10)**

The same business rule — *a failed `UPLOADING_VIDEO` or `UPDATING_METADATA` stage means the run is NOT a clean success* — is independently re-encoded in two places:

- `src/ui/src/hooks/useShaveManager.ts` — `hasRequiredStageFailed()` downgrades the **persisted** status `ShaveStatus.Completed -> Failed`:
  ```ts
  return state.uploading_video.status === "failed" || state.updating_metadata.status === "failed";
  ```
- `src/ui/src/components/workflow/FinalResultPanel.tsx` — `deriveStageWarning()` / `resolveDisplayStatus()` downgrade the **on-screen badge** to `"warning"`, checking the same two stages via the `WorkflowProgressStage` enum.

These encode the identical rule over the identical stage set (`UPLOADING_VIDEO === "uploading_video"`, `UPDATING_METADATA === "updating_metadata"`), just expressed two different ways — raw lowercase keys vs enum constants. The "required post-creation stages" set is hardcoded in both.

If a stage is added, renamed, or removed in one function but not the other, the persisted shave status and the on-screen badge will disagree — exactly the "status lies about what happened" class of bug this PR exists to kill. The copies currently agree, so this is a latent drift risk rather than an active bug.

AGENTS.md Rule 10 (DRY): *"If the same logic appears in more than one place, extract it into a shared function, hook, or utility. Duplicated code leads to bugs when one copy is updated but the other is forgotten."*

**Remedy:** Extract one shared predicate/helper (e.g. `requiredPostCreationStageFailure(state)` returning the failed stage + its error) into a shared location both files import — per Rule 9 a frontend cross-layer helper belongs under `src/ui/src/utils` or `src/ui/src/types`. Have both the persisted-status downgrade and the warning badge derive from it. The two distinct behaviors (boolean for persistence, message for display) can both build on the single predicate, so no over-abstraction is introduced.

File: src/ui/src/hooks/useShaveManager.ts [hunk 3]
```diff
@@ -53,6 +53,17 @@ function parseFinalOutput(finalOutput: string): ParsedShaveOutput | null {
...
 
+/**
+ * #861 AC#3/#5: the run must NOT be persisted as successfully Completed when a required
+ * post-creation stage (video upload OR YouTube metadata update) failed. The AI's Status
+ * still reads "success" in that case, so we downgrade the persisted status to Failed here.
+ * The work item URL is still saved so the user can reach it — the status just tells the
+ * truth that not every required step completed without errors.
+ */
+function hasRequiredStageFailed(state: WorkflowState): boolean {
+  return state.uploading_video.status === "failed" || state.updating_metadata.status === "failed";
```
Comment (calumjs) [inline src/ui/src/hooks/useShaveManager.ts]: DRY / AGENTS.md Rule 10: this same "a failed required post-creation stage means not a clean success" rule is also encoded in `FinalResultPanel.tsx` (`deriveStageWarning()`/`resolveDisplayStatus()`) over the identical stage set, expressed via the `WorkflowProgressStage` enum. The two copies will drift if a required stage is added/renamed/removed in one but not the other — the persisted status and the on-screen badge would then disagree. Extract a single shared predicate (e.g. `requiredPostCreationStageFailure(state)`) under `src/ui/src/utils` that both derive from.

Comment (calumjs) [reply src/ui/src/hooks/useShaveManager.ts]: Fixed in d300dcd. Extracted a single source of truth, `requiredPostCreationStageFailure(state)`, into `src/ui/src/utils/index.ts`. It returns `{ stage, error } | null` — the first failed required post-creation stage (UPLOADING_VIDEO, then UPDATING_METADATA) plus a human-readable message (the stage's own payload `error` when present, otherwise a sensible default).

Both call sites now derive from it, so the persisted status and the on-screen badge can no longer drift:
- `useShaveManager` downgrades `ShaveStatus.Completed -> Failed` when `requiredPostCreationStageFailure(state)` is non-null.
- `FinalResultPanel.deriveStageWarning()` is now just `requiredPostCreationStageFailure(state)?.error ?? null`, feeding `resolveDisplayStatus()` as before.

The stage set + messages are defined once (`REQUIRED_POST_CREATION_STAGE_MESSAGES`), so adding/renaming/removing a required stage is a single edit. Added unit tests in `index.test.ts` covering: no failure, each stage's default message, payload-error precedence over the default, and upload-takes-precedence-over-metadata.

Validation: root `npm run build` (tsc) clean; the new `requiredPostCreationStageFailure` tests pass (13/13 in the utils suite under the UI vite alias); `biome check` clean on all four changed files.

**Review by calumjs — COMMENTED** [review]
## Review summary

One material correctness finding that inverts this PR's own stated AC5 for issue #861.

### [Medium] Successful metadata-only retry leaves the shave permanently marked `Failed` (AC5 violation)

`src/ui/src/hooks/useShaveManager.ts:223`

The dedupe key is only `${shaveId}:${finalOutput}`, and the hook early-returns once that key is seen:

```ts
const finalUpdatedKey = `${shaveId}:${finalOutput}`;
if (finalUpdatedKeysRef.current.has(finalUpdatedKey)) { return; }
```

The new #861 `shaveStatus` downgrade logic sits **inside** that guarded block. Trigger:

1. First run emits `executing_task.completed` (final output present) **plus** `updating_metadata.failed`, so `requiredPostCreationStageFailure(state)` downgrades the persisted shave to `Failed` and the key `${shaveId}:${finalOutput}` is recorded.
2. The user retries **only** `UPDATING_METADATA`. `prepareStageForRetry('updating_metadata')` keeps `executing_task='completed'` and resets metadata onward; `retryFromMiddleStage -> processVideoSource(startFromStage='updating_metadata')` makes `shouldRunStage(EXECUTING_TASK)` false, so `executing_task` is not re-run and its checkpoint `finalOutput` is reused unchanged.
3. On metadata success, `completeStage(UPDATING_METADATA)` broadcasts `WORKFLOW_PROGRESS_NEO` with `executing_task=completed` + `updating_metadata=completed`. `isWorkflowReadyForFinalOutput` is true and `finalOutput` is the **same string**, so the dedupe key matches and the hook returns early.

`requiredPostCreationStageFailure(state)` would now return `null` (all required stages complete), but the recompute is short-circuited, so the persisted `shaveStatus` stays `Failed`. No backend path rewrites `shaveStatus` on retry (`retryFromStage` returns without touching it), so nothing corrects it.

This contradicts #861 AC5: *"A bug report is only marked as successfully processed when all required steps complete without errors"* — here all required steps complete, yet the report stays `Failed`.

**Remedy:** include the relevant post-creation stage status in the dedupe key (or re-run when `requiredPostCreationStageFailure(state)` changes) so a successful metadata retry can upgrade the persisted status once all required stages complete.

File: src/ui/src/hooks/useShaveManager.ts [hunk 4]
```diff
@@ -231,9 +232,18 @@ export function useShaveManager() {
       if (parsedOutput) {
         const finalTitle = parsedOutput.title || uploadResult?.data?.title || "Untitled Work Item";
 
+        // #861 AC#3/#5: a failed required post-creation stage overrides the AI's "success"
+        // so the persisted shave status doesn't claim a clean completion. This shares the
+        // single source of truth with FinalResultPanel's warning badge, so the persisted
+        // status and the on-screen badge can never disagree.
+        const shaveStatus =
+          parsedOutput.status === ShaveStatus.Completed && requiredPostCreationStageFailure(state)
```
Comment (calumjs) [inline src/ui/src/hooks/useShaveManager.ts]: [Medium] Dedupe key `${shaveId}:${finalOutput}` short-circuits the #861 status recompute, so a successful metadata-only retry can't upgrade a previously-`Failed` shave.

A first run with `executing_task.completed` + `updating_metadata.failed` downgrades the persisted shave to `Failed` and records this key. A subsequent `UPDATING_METADATA`-only retry reuses the same `finalOutput` (executing_task isn't re-run), so on success the identical key matches and the hook returns early before `requiredPostCreationStageFailure(state)` is re-evaluated. The shave stays `Failed` even though all required stages now completed — violating #861 AC5.

Fix: include the post-creation stage status in the dedupe key (or re-run when `requiredPostCreationStageFailure(state)` changes).

Comment (calumjs) [reply src/ui/src/hooks/useShaveManager.ts]: Fixed in 84754dd. The dedupe key now incorporates the required post-creation stage failure: `${shaveId}:${finalOutput}:${postCreationFailure?.stage ?? "none"}`. On the first run the key carries the failed stage (e.g. `updating_metadata`); after a metadata-only retry succeeds, `requiredPostCreationStageFailure(state)` returns `null`, so the key changes to `…:none`, the early-return no longer matches, and the hook re-runs the status recompute — upgrading the persisted shave from `Failed` to the AI-reported success once all required stages complete (#861 AC5). The single `postCreationFailure` value is reused for the status decision so the key and the persisted status stay consistent.

Validation: root `tsc` build + UI `tsc --noEmit` pass; non-DB vitest suite passes (212 tests, incl. `index.test.ts` covering `requiredPostCreationStageFailure`); `biome check` clean on the changed file. (DB-service and ffmpeg-native suites were skipped — they need native modules unavailable in this env, unrelated to this change.)

**Review by tomek-i — APPROVED** [review]

**Review by tomek-i — APPROVED** [review]
